drivers/regulator/regulator_npm6001.c

```diff
@@ -0,0 +1,690 @@
+	const struct device *dev, int32_t min_uv, int32_t max_uv,
+	const struct linear_range *range, uint8_t vout_reg, uint8_t conf_reg)
+{
+	const struct regulator_npm6001_config *config = dev->config;
+	const struct regulator_npm6001_pconfig *pconfig = config->p->config;
+	uint8_t conf, buf[3];
+	uint16_t idx;
+	int ret;
+
+	ret = linear_range_get_win_index(range, min_uv, max_uv, &idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* force PWM mode when updating voltage */
+	ret = i2c_reg_read_byte_dt(&pconfig->i2c, conf_reg, &conf);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if ((conf & NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM) == 0U) {
+		ret = i2c_reg_write_byte_dt(
+			&pconfig->i2c, conf_reg,
+			conf | NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	/* write voltage in both ULP/PWM registers */
+	buf[0] = vout_reg;
+	buf[1] = (uint8_t)idx;
+	buf[2] = (uint8_t)idx;
+
+	ret = i2c_write_dt(&pconfig->i2c, buf, sizeof(buf));
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = i2c_reg_write_byte_dt(&pconfig->i2c, NPM6001_TASKS_UPDATE_VOUTPWM,
+				    1U);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* restore HYS mode if enabled before */
+	if ((conf & NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM) == 0U) {
+		ret = i2c_reg_write_byte_dt(&pconfig->i2c, conf_reg, conf);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int regulator_npm6001_buck3_set_voltage(const struct device *dev,
+					       int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_npm6001_config *config = dev->config;
+	const struct regulator_npm6001_pconfig *pconfig = config->p->config;
+	uint16_t idx;
+	uint8_t conf;
+	int ret;
+
+	ret = linear_range_get_win_index(&buck3_range, min_uv, max_uv, &idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* force PWM mode when updating voltage */
+	ret = i2c_reg_read_byte_dt(&pconfig->i2c, NPM6001_BUCK3CONFPWMMODE,
+				   &conf);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if ((conf & NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM) == 0U) {
+		ret = i2c_reg_write_byte_dt(
+			&pconfig->i2c, NPM6001_BUCK3CONFPWMMODE,
+			conf | NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	ret = i2c_reg_write_byte_dt(&pconfig->i2c, NPM6001_BUCK3VOUT,
+				    (uint8_t)idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* restore HYS mode if enabled before */
+	if ((conf & NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM) == 0U) {
+		ret = i2c_reg_write_byte_dt(&pconfig->i2c,
+					    NPM6001_BUCK3CONFPWMMODE, conf);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int regulator_npm6001_ldo0_set_voltage(const struct device *dev,
+					      int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_npm6001_config *config = dev->config;
+	const struct regulator_npm6001_pconfig *pconfig = config->p->config;
+	uint8_t val;
+
+	if ((min_uv < 1800000) || (max_uv > 3300000)) {
+		return -EINVAL;
+	}
+
+	if ((min_uv > 3000000) && (max_uv <= 3300000)) {
+		val = NPM6001_LDO0VOUT_SET3V3;
```

Comment:
Perhaps I'm missing something, but I got confused trying to understand how is this actually supposed to work.
Let's say that a user calls `regulator_set_voltage()` with `min_uv = 1500000` and `max_uv = 3600000`, because the acceptable voltage is anything between 1.5 and 3.6 V. Why would this function return `-EINVAL` then?
On the other hand, when this function gets called with, for example, `min_uv = 3100000` and `max_uv = 3200000`, it will configure the voltage to be 3.3 V, so beyond the specified range. Is that okay?